sw/host/opentitanlib/src/backend/mod.rs

```diff
@@ -66,65 +68,69 @@ pub enum Error {
-                args.usb_serial.as_deref(),
-            )?);
-            (dediprog, Some(Path::new("/__builtin__/dediprog.json")))
-        }
-        _ => return Err(Error::UnknownInterface(interface.to_string()).into()),
-    };
     if args.conf.is_empty() {
-        if let Some(conf_file) = default_conf {
-            process_config_file(&mut env, conf_file)?
+        if let Some(conf_file) = get_config_file(interface)? {
+            process_config_file(&mut builder, Path::new(conf_file))?
         }
     }
-    env.build(backend)
+    let interface = builder.get_interface();
+    let io_mapper = Rc::new(builder.build_io_mapper()?);
+    let backend = build_transport(interface, Rc::clone(&io_mapper), args)?;
+    builder.build_wrapper(backend, io_mapper)
+}
+
+fn build_transport(
+    interface: &str,
+    io_mapper: Rc<IoMapper>,
+    args: &BackendOpts,
+) -> Result<Box<dyn crate::transport::Transport>> {
+    Ok(match interface {
+        "" => create_empty_transport()?,
+        "proxy" => proxy::create(&args.proxy_opts)?,
+        "verilator" => verilator::create(&args.verilator_opts, io_mapper)?,
+        "ti50emulator" => ti50emulator::create(&args.ti50emulator_opts)?,
+        "ultradebug" => ultradebug::create(args, io_mapper)?,
+        "hyper310" => hyperdebug::create::<CW310Flavor>(args)?,
+        "hyperdebug" => hyperdebug::create::<StandardFlavor>(args)?,
+        "hyperdebug_dfu" => hyperdebug::create_dfu(args)?,
+        "c2d2" => hyperdebug::create::<C2d2Flavor>(args)?,
+        "servo_micro" => hyperdebug::create::<ServoMicroFlavor>(args)?,
+        "ti50" => hyperdebug::create::<Ti50Flavor>(args)?,
+        "cw310" => chip_whisperer::create::<Cw310>(args, io_mapper)?,
+        "cw340" => chip_whisperer::create::<Cw340>(args, io_mapper)?,
+        "dediprog" => Box::new(Dediprog::new(
+            args.usb_vid,
+            args.usb_pid,
+            args.usb_serial.as_deref(),
+            io_mapper,
+        )?),
+        _ => return Err(Error::UnknownInterface(interface.to_string()).into()),
+    })
+}
+
+fn get_config_file(interface: &str) -> Result<Option<&str>> {
+    Ok(Some(match interface {
+        "" => return Ok(None),
+        "proxy" => return Ok(None),
+        "hyperdebug" => return Ok(None),
+        "hyperdebug_dfu" => return Ok(None),
+        "ti50" => return Ok(None),
+        "verilator" => "/__builtin__/opentitan_verilator.json",
+        "ti50emulator" => "/__builtin__/ti50emulator.json",
+        "ultradebug" => "/__builtin__/opentitan_ultradebug.json",
+        "hyper310" => "/__builtin__/hyperdebug_cw310.json",
+        "c2d2" => "/__builtin__/h1dx_devboard_c2d2.json",
+        "servo_micro" => "/__builtin__/servo_micro.json",
+        "cw310" => "/__builtin__/opentitan_cw310.json",
+        "cw340" => "/__builtin__/opentitan_cw340.json",
+        "dediprog" => "/__builtin__/dediprog.json",
+        _ => return Err(Error::UnknownInterface(interface.to_string()).into()),
+    }))
 }
```

Comment:
Repeating these match statements feels unnecessary. A `build_transport_and_get_config` function that returns tuples (similar to the original code) may be nicer.

Reply:
I had to split these functions because the transport requires a `IoMapper` which requires a `config.json`, so these must be separated methods.